---
title: "Exome-seek/dragen mixed workflow with mouse samples"
author: "Bingfang Xu"
date: "3/6/2022"
output:
  html_document:
    df_print: paged
  word_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r echo=FALSE, warning=FALSE, message=FALSE}
library(tidyverse)
#install.packages("officer")
library(officer)
#install.packages("flextable")
library(flextable)
library(dplyr)

```

# Sample information

Use 12_RPZ0113 (Highseq, Jan, 2019) or 1_RPZ27911 (Novaseq, sep,2020) as normal.

```{r echo=FALSE, warning=FALSE, message=FALSE}

library(tidyverse)
samples <-  read_tsv("/Users/xubr/local_projects/20230208_10MEF/Reports/pairs.txt") 
knitr::kable(samples)

```

How to untar the files

```         
tar -xvf yourfile.tar  ## extract the file to the current directory.
tar -C /myfolder -xvf yourfile.tar  ##extract to another directory.
```

# Exome-seek germline calling

Use snp_indel.filtered.vcf.gz as input.

### Filter variants of germline calling, then VCF to VEP on Biowulf.

```         
### Remove common mouse germline SNPs on biowulf
sinteractive --mem 32g
module load GATK/3.8-1 # current vertion GATK/4.2
java -Xmx8g -jar $GATK_JAR -T SelectVariants -R /data/xubr/mm10_ordered_all.fa -V /data/xubr/snp_indel.filtered.vcf.gz --discordance /data/xubr/sorted2_mgp.v4.all.dbSNP_Sample_129S5SvEvBrd.vcf -o joint_noncommonSNP.vcf -U ALLOW_SEQ_DICT_INCOMPATIBILITY

### Remove variants already existed in RPZ113 using data from three PRZ113 sequencing(200ngRPZ113, 3ngRPZ113, and RPZ0113)
java -Xmx8g -jar $GATK_JAR -T SelectVariants -R /data/xubr/mm10_ordered_all.fa -V /data/xubr/joint_noncommonSNP.vcf --discordance /data/xubr/combined_GATK_3RPZ113.vcf -o /data/xubr/joint_noncommonSNP_nonRPZ113.vcf -U ALLOW_SEQ_DICT_INCOMPATIBILITY

### Remove non-exon variants
java -Xmx8g -jar $GATK_JAR -T SelectVariants -R mm10_ordered_all.fa -L S0276129_Regions_mm10.bed --variant joint_noncommonSNP_nonRPZ113.vcf -o Ontarget_joint_noncommonSNP_nonRPZ113.vcf

### VCF to VEP
module load VEP/100
vep -i ./Ontarget_joint_noncommonSNP_nonRPZ113.vcf -o Ontarget_joint_noncommonSNP_nonRPZ113_vep.txt --offline --cache --dir_cache $VEP_CACHEDIR --species mouse --fasta $VEP_CACHEDIR/mouse.fa --everything --af_exac --vcf

exit

### move files (vcf and vef files) to local.
```

### Generate variant lists of germline calling

output file "germline_total.txt"

```         
### Filter by impact (high, moderate and deleterious), AF>0.2, and DP>30. 
### Extract information of AF, IMPACT, MUTATION, Amino acid change. 
### Remove a list of misaligned genes. 
### Sort variants by gene name.
python ./_1filter_VEP_gemlineList.py
```

# Dragen somatic calling

### Filter variants of somatic calling

The maf file with black list gene removed were generated by dragen pipeline "somatic_blackListGenesRemovedCombined.maf.txt"

```         
# input maf file somatic_blackListGenesRemovedCombined.maf or somatic_combined.maf
# mouse black list is a empty list
maf <- read.delim("/Users/xubr/local_projects/20230208_10MEF/dragen_somatic/preCombine/somatic_combined.maf", skip =1, header = TRUE, sep = "\t")
colnames(maf)


# simplify maf file, remove 'p.' from HGVSp, save as "*_selectedColumn.txt"
narrow_maf <- maf %>%
  # remove extra columns
  select(c(`Hugo_Symbol`,`Chromosome`, `Start_Position`, `Variant_Classification`,`Tumor_Sample_Barcode`,`HGVSp`,`t_depth`,`t_ref_count`,`t_alt_count`,`SIFT`,`IMPACT`,`VARIANT_CLASS`,`FILTER`))
narrow_maf$`HGVSp`=substring(narrow_maf$`HGVSp`, first=3)               

write.table(narrow_maf, "/Users/xubr/local_projects/20230208_10MEF/dragen_somatic/preCombine/somatic_Combined_selectedColumn.txt", sep = "\t",row.names = FALSE, col.names = TRUE)


# filter by IMPACT,FILTER, AF, t_depth,classify variants, and sort, save as"*_selectedColumn_filtered.txt"
short_maf <- narrow_maf %>%
  filter(`IMPACT`=="HIGH"|`IMPACT`=="MODERATE") %>%
  filter(!str_detect(`SIFT`,"tolerated") ) %>%
  filter(`FILTER`=="PASS") %>%
  mutate(AF=round((`t_alt_count`/`t_depth`), digits = 2)) %>%
  filter(`AF`>0.2 & t_depth > 30) %>%
  mutate(HGVSp=if_else(HGVSp=='',"Splicing variant", HGVSp)) %>%
  mutate(CLASS=case_when(AF > 0.9 | AF == 0.9 ~ "I-A",
                       AF > 0.2 & AF < 0.9 & `IMPACT`=="HIGH" ~ "I-B",
                       AF > 0.2 & AF < 0.9 & `IMPACT`=="MODERATE" ~ "II-A")) %>%
  mutate(CLASS=if_else(str_detect(HGVSp,"(Ala[0-9]*Gly$)|(Gly[0-9]*Ala$)|(Ser[0-9]*Thr$)|(Thr[0-9]*Ser$)|(Ile[0-9]*Leu$)|(Leu[0-9]*Ile$)"),"II-B", CLASS)) %>%
  arrange(`IMPACT`,`Hugo_Symbol`,`Chromosome`, `Start_Position`) 
                         
write.table(short_maf, "/Users/xubr/local_projects/20230208_10MEF/dragen_somatic/preCombine/somatic_combined_selectedColumn_filtered.txt", sep = "\t",row.names = FALSE, col.names = TRUE)
```

### Export individual somatic variant file for for each sample

```         
# prepare for individual file for each sample
final_maf<- short_maf  %>%
  select(c(`Hugo_Symbol`,`Chromosome`, `Start_Position`,`AF`,`IMPACT`,`HGVSp`,`CLASS`,`Tumor_Sample_Barcode`,`VARIANT_CLASS`))%>%
  arrange(`IMPACT`,`Hugo_Symbol`)   
colnames(final_maf) = c('GENE','CHR','POS','AF','IMPACT','AA CHANGES','CLASS','Tumor_Sample_Barcode','VARIANT_CLASS')
maf_group <- final_maf %>% group_split(`Tumor_Sample_Barcode`)
colnames((maf_group[[1]]))


for (i in c(1:10)){
  file_name=paste0(maf_group[[i]]$`Tumor_Sample_Barcode`[1],"_annotated.txt",collapse ="")
  print(file_name)
  sample1 <- maf_group[[i]] %>% select(-c(`Tumor_Sample_Barcode`))
  #colnames(sample)
  write.table(sample1, paste0("/Users/xubr/local_projects/20230208_10MEF/dragen_somatic/preCombine/", file_name), sep = "\t",row.names = FALSE, col.names = TRUE)

}
```

# Combine variants from germline and somatic calling

```         
### input two files: "germline_hardFiltered_total.txt" and "somatic_Combined_selectedColumn_filtered.txt"
### Add a "caller" column to annotate the methods which call the variant (germline, somatic, or both) 
### Add "sample_name" column.
### Add 3 column: "Notes", "Add_to_somatic_list","Remove_from_somatic_list"    
./_2preIGV.bat
python _2compare_marked.py
python _3addSampleName.py 


########### Manually IGV check.##########

### 1. Generate a summary excel file.
### 2. Do IGV check manually.  Focus on varaints with flag "check".
### 4. Save as a txt file "IGV_Final.txt", choose tab delimited.

############ Merge anootated data frame #############################
IGV1 <- read.delim("/Users/xubr/local_projects/20230208_10MEF/dragen_somatic/combined_with_germline/combined_germline_somatic_marked_addSampleName.txt", header = TRUE, sep = "\t")

IGV_1 <- IGV1 %>% select(-c(`Notes`,`Add_to_somatic_list`,`Remove_from_somatic_list`))

colnames(IGV_1)

IGV2 <- read.delim("/Users/xubr/local_projects/20230208_10MEF/dragen_somatic/combined_with_germline/Final_IGV.txt", header = TRUE, sep = "\t")

IGV_2 <- IGV2 %>% select(c(`GENE`,  `X.CHROM`,  `POS`,`Notes`, `Add_to_somatic_list`,`Remove_from_somatic_list`))

colnames(IGV_2)
IGV4<- distinct(left_join(x=IGV_1,y=IGV_2, by = c("GENE","X.CHROM","POS")))



write.table(IGV4, "/Users/xubr/local_projects/20230208_10MEF/dragen_somatic/combined_with_germline/Final_IGV_test.txt", sep = "\t",row.names = FALSE, col.names = TRUE)
##############


  
### Add real germline variants to somatic list for individual samples.
### Mark wrong somatic variants with "remove".
### Recalculate AF for the variants from germline calling.
### Define class
./_3postIGV.bat
python _5add_minus_additional.py
python _6defineClass.py 
python _7reShape.py 
```

# Write report

```         

setwd("~/local_projects/20230208_10MEF/dragen_somatic/combined_with_germline")

data_files <- list.files("/Users/xubr/local_projects/20230208_10MEF/dragen_somatic/combined_with_germline/", pattern = "*_annotated_additional_Final_fixed.txt")  

for (ff in data_files) {
   base <- str_split(ff, "_annotated" )
   print(base)

   maf_temp <- read.delim(paste0("/Users/xubr/local_projects/20230208_10MEF/dragen_somatic/combined_with_germline/",ff), header = TRUE, sep = "\t")
   
   # obtain sample name, file name, date
   sample_name <- base[[1]][1]
   print(sample_name)
   file_name=paste0(base[[1]][1],".docx",collapse ="")

   today <- Sys.Date()
   format(today, format="%B %d %Y")
   print(today)
   # open an empty docx file, write sample name
   doc = read_docx()
   prop1=fp_text(font.size = 12, bold = TRUE, font.family = "Times")
   title1=fpar(ftext(sample_name, prop=prop1))
   body_add_fpar(doc, title1)
 
   # write subtitle and date
   body_add_par(doc, "Whole Exome report ")
   body_add_par(doc, today)
   body_add_par(doc, " ")

   # write method summary
   body_add_par(doc, "Method Summary: ")
   body_add_par(doc, "DNA was extracted using Qiagen all prep protocol. gDNA libraries from 500ng DNA was prepared and exons were captured using Agilent SureselectXT mouse all exon protocol. Libraries were quantitated via Agilent TapeStation. Sequencing was performed on the Illumina NextSeq2000. Raw FastQ files were mapped using the Dragen v3.10 and exome-seek v1.0.3-beta protocols for mutation detection. Single Nucleotide variants and insertions or deletions were identified, and a subset of calls filtered by quality, depth of coverage, and allele frequency were compiled into a variant list. A subset of the compiled variants was reviewed manually using the Integrated Genome Viewer software from the Broad Institute. ")
   body_add_par(doc, " ")
   
   body_add_par(doc, "FastQ and BAM files are stored at smb://at-s-is2.ncifcrf.gov/ras-intl/static/Genomics/20230225_10samples_diploidMEF/")
   body_add_par(doc, " ")
 
  # insert a table
  body_add_par(doc, "Summary of findings: ")
  count_impact <- maf_temp %>% dplyr::count(IMPACT,VARIANT) 
  

  f.table=qflextable(count_impact)
  f.table=font(f.table,  fontname = "Times", part = "all")
  f.table=fontsize(f.table, size = 10, part = "all")
  doc <- flextable::body_add_flextable(doc, value = f.table, align = "left" )
  body_add_par(doc, " ")
 

  

  # insert gene list
  body_add_par(doc, "Gene list: ")
  maf1 <- maf_temp %>% select(-c(VARIANT))
  f.table=qflextable(maf1)
  set_table_properties(f.table, width = .5, layout = "autofit")
  f.table=font(f.table,  fontname = "Times", part = "all")
  f.table=fontsize(f.table, size = 10, part = "all")
  # also set the table's header font as bold
  f.table=bold(f.table, part = "header")
  doc <- flextable::body_add_flextable(doc, 
                                       value = f.table, 
                                       align = "center" )
  body_add_par(doc, " ")
  body_add_par(doc, " ")
  body_add_par(doc, "I-A    Mutation call is well supported and there is a high probability that the variant will impact RAS dependent proliferation in this cell (e.g. Homozygous mutation of Trp53).")
  body_add_par(doc, " ")
  body_add_par(doc, "Variants called by one of variant calling methods and visualized by IGV. Single variant on a gene, AF>0.9 with high impact. Or multiple variants on a gene, 0.9>AF>0.2 with high impact.")
  body_add_par(doc, " ")
  body_add_par(doc, "Variants called by one of variant calling methods and visualized by IGV. Single variant on a gene, AF>0.9 with moderate impact. Or multiple variants on a gene, 0.9>AF>0.2 with moderate impact.")
  body_add_par(doc, " ")
  body_add_par(doc, "I-B    Mutation call is well supported, the variant is probably damaging, the variant may impact cell growth or genetic stability.")
  body_add_par(doc, " ")
  body_add_par(doc, "Variants called by one of variant calling methods and visualized by IGV. Single variant on a gene, 0.9>AF>0.2 with high impact. ")
  body_add_par(doc, " ")
  body_add_par(doc, " II-A    Call is well supported, variant is non-synonymous and likely impact protein function.") 
  body_add_par(doc, " ")               
  body_add_par(doc, "Variants called by one of variant calling methods and visualized by IGV. Single variant on a gene, 0.9>AF>0.2 with moderate impact.") 
  body_add_par(doc, " ")
  body_add_par(doc, "II-B    Call is well supported, non-synonymous change in amino acids with minor change in amino acid properties (e.g.  Serine to Threonine, Ile to Leu, or Glycine to Alanine) that may alter protein function.")
  body_add_par(doc, " ")
  body_add_par(doc, "Variants called by one of variant calling methods and visualized by IGV. Single variant on a gene, AF>0.2 with moderate impact and with minor change in amino acid properties (e.g.  Serine to Threonine, Ile to Leu, or Glycine to Alanine). ")
  body_add_par(doc, " ")
  body_add_par(doc, "II-C    Mutation call is well supported but the change is unlikely to impact protein function.")

  
  # output docx file
  print(doc, target=file_name)
  
}
```
